binding.go

```diff
@@ -201,13 +204,28 @@ func Json(jsonStruct interface{}, ifacePtr ...interface{}) macaron.Handler {
 		var errors Errors
 		ensureNotPointer(jsonStruct)
 		jsonStruct := reflect.New(reflect.TypeOf(jsonStruct))
-		if ctx.Req.Request.Body != nil {
-			defer ctx.Req.Request.Body.Close()
-			err := json.NewDecoder(ctx.Req.Request.Body).Decode(jsonStruct.Interface())
-			if err != nil && err != io.EOF {
-				errors.Add([]string{}, ERR_DESERIALIZATION, err.Error())
+		var err error
+		if ctx.Req.Method == "POST" || ctx.Req.Method == "PUT" || ctx.Req.Method == "PATCH" {
+			if ctx.Req.Request.Body != nil {
+				defer ctx.Req.Request.Body.Close()
+				v := jsonStruct.Interface()
+				if pb, ok := v.(proto.Message); ok {
+					err = jsonpb.Unmarshal(ctx.Req.Request.Body, pb)
+				} else {
+					err = json.NewDecoder(ctx.Req.Request.Body).Decode(v)
+				}
 			}
 		}
+		if (err == nil || err == io.EOF) && ctx.Req.URL != nil {
+			if params := ctx.Req.URL.Query(); len(params) > 0 {
+				d := schema.NewDecoder()
+				d.SetAliasTag("json")
+				err = d.Decode(jsonStruct.Interface(), params)
+			}
```

Comment:
What is the purpose of having line 220-223?

Reply:
This brings `JSON()` method parity with `Form()` method. `Form()` method parses request body for `POST/PUT/PATCH` and query parameters for all request.
https://github.com/golang/go/blob/master/src/net/http/request.go#L1194

`JSON()` currently parses request body for any HTTP method. This pr brings `JSON()` method parity to `Form()`. LN 220-223 will parse query parameters into struct using the json tags to match with the query parameter keys.

Reply:
@tamalsaha I see, but I think in this package JSON is JSON, it is meant to be presented in the body. Otherwise, we should provide an option to enable this automatic query parameter to struct.

Reply:
For `POST/PUT/PATCH` methods, request parameters are sent via request body. But for GET method, the request parameters always coming via query parameter. 

To me it should be ok to do this automatically. But if we are to add options, what should that be?